src/telemetry/mod.rs

```diff
@@ -0,0 +1,345 @@
+					exporter.version.clone(),
+				)?;
+
+				// Create a new OpenTelemetryTracingBridge using the above LoggerProvider.
+				let logs_layer = OpenTelemetryTracingBridge::new(&logger_provider);
+				let logs_layer = logs_layer.with_filter(filter_otel);
+
+				Ok::<_, OtelInitError>((Some(logger_provider), Some(logs_layer)))
+			})
+		})
+		.transpose()?
+		.unwrap_or((None, None));
+
+	let (tracer_provider, tracer_layer) = config
+		.exporter
+		.as_ref()
+		.and_then(|exporter| {
+			exporter.traces.as_ref().and_then(|c| c.enabled.then_some(c)).map(|tracer_config| {
+				let trace_filter = EnvFilter::from_str(&tracer_config.get_filter(main_crate))?;
+				let tracer_provider = init_traces(
+					exporter.get_endpoint(),
+					exporter.service_name.clone(),
+					exporter.version.clone(),
+				)?;
+				let tracer = tracer_provider.tracer(exporter.service_name.clone());
+				let tracer_layer = OpenTelemetryLayer::new(tracer).with_filter(trace_filter);
+				Ok::<_, OtelInitError>((Some(tracer_provider), Some(tracer_layer)))
+			})
+		})
+		.transpose()?
+		.unwrap_or((None, None));
+
+	let (meter_provider, meter_layer) = config
+		.exporter
+		.as_ref()
+		.and_then(|exporter| {
+			exporter.metrics.as_ref().and_then(|c| c.enabled.then_some(c)).map(|meter_config| {
+				let metrics_filter = EnvFilter::from_str(&meter_config.get_filter(main_crate))?;
+				let meter_provider = init_metrics(
+					exporter.get_endpoint(),
+					exporter.service_name.clone(),
+					exporter.version.clone(),
+				)?;
+				let meter_layer =
+					MetricsLayer::new(meter_provider.clone()).with_filter(metrics_filter);
+
+				Ok::<_, OtelInitError>((Some(meter_provider), Some(meter_layer)))
+			})
+		})
+		.transpose()?
+		.unwrap_or((None, None));
+
+	// Initialize the tracing subscriber with the stdout layer and
+	// layers for exporting over OpenTelemetry the logs, traces and metrics.
+	tracing_subscriber::registry()
+		.with(logs_layer)
+		.with(stdout_layer)
+		.with(meter_layer)
+		.with(tracer_layer)
+		.init();
+
+	Ok(ProvidersGuard { logger_provider, tracer_provider, meter_provider })
+}
+
+/// Guarding object to make sure the providers are properly shutdown
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct ProvidersGuard {
+	logger_provider: Option<LoggerProvider>,
+	tracer_provider: Option<TracerProvider>,
+	meter_provider: Option<SdkMeterProvider>,
+}
+
+// Necessary to call TracerProvider::shutdown() on exit
+// due to a bug with flushing on global shutdown:
+// https://github.com/open-telemetry/opentelemetry-rust/issues/1961
+impl Drop for ProvidersGuard {
+	fn drop(&mut self) {
+		// This causes a hang in testing.
+		// Some relevant information:
+		// https://github.com/open-telemetry/opentelemetry-rust/issues/536
+		#[cfg(not(test))]
+		{
+			self.logger_provider.as_ref().inspect(|logger_provider| {
+				if let Err(err) = logger_provider.shutdown() {
+					tracing::error!("Could not shutdown LoggerProvider: {err}");
+				}
+			});
```

Comment:
IMHO: I think it makes more sense for `if let` syntax on `self` fields as it's more about flow control, while `inspect` and `inspect_err` semantially are more about "inspection", about some side effects that are not part of general flow, like tracing. Feel free to disagree, I would just write it like this:
```suggestion
			if let Some(logger_provider) = self.logger_provider.as_ref() {
				logger_provider.shutdown().inspect_err(|err|
					tracing::error!("Could not shutdown LoggerProvider: {err}")
				)
			};
```